src/backends/perf_pt/mod.rs

```diff
@@ -0,0 +1,221 @@
+    aux_bufsize: size_t,
+    /// Opaque C pointer representing the tracer context.
+    tracer_ctx: Option<*const c_void>,
+}
+
+impl PerfPTTracer {
+    /// Create a new tracer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use traceme::backends::PerfPTTracer;
+    ///
+    /// PerfPTTracer::new().trace_filename("mytrace.ptt").data_bufsize(1024).target_pid(666);
+    /// ```
+    pub fn new() -> Self {
+        Self {
+            trace_filename: String::from("traceme.ptt"),
+            target_pid: unsafe { getpid() },
+            tracer_ctx: None,
+            data_bufsize: 64,
+            aux_bufsize: 1024,
+        }
+    }
+
+    /// Set where to write the trace.
+    ///
+    /// # Arguments
+    ///
+    /// * `filename` - The filename in which to store trace packets.
+    pub fn trace_filename(mut self, filename: &str) -> Self {
+        self.trace_filename = String::from(filename);
+        self
+    }
+
+    /// Select which PID to trace.
+    ///
+    /// By default, the current PID is traced.
+    ///
+    /// # Arguments
+    ///
+    /// * `pid` - The PID to trace.
+    pub fn target_pid(mut self, pid: pid_t) -> Self {
+        self.target_pid = pid;
+        self
+    }
+
+    /// Set the PT data buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The data buffer size to use.
+    pub fn data_bufsize(mut self, size: usize) -> Self {
+        self.data_bufsize = size;
+        self
+    }
+
+    /// Set the PT aux buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The aux buffer size to use.
+    pub fn aux_bufsize(mut self, size: usize) -> Self {
+        self.aux_bufsize = size;
+        self
+    }
+
+    /// Make the map filename by setting/adding a ".map" extension to `trace_filename`.
+    fn make_map_filename(trace_filename: &str) -> Result<String, TraceMeError> {
+        let mut pb = PathBuf::from(trace_filename);
+        if !pb.set_extension("map") {
+            return Err(TraceMeError::InvalidFileName("".to_string()));
+        }
+        let map_filename = pb.to_str().ok_or(TraceMeError::InvalidFileName("".to_string()))?;
+        Ok(String::from(map_filename))
+    }
+
+    fn check_perf_perms() -> Result<(), TraceMeError> {
+        if unsafe { geteuid() } == 0 {
+            // Root can always trace.
+            return Ok(());
+        }
+
+        let mut f = File::open(&PERF_PERMS_PATH)?;
+        let mut buf = String::new();
+        f.read_to_string(&mut buf)?;
+        let perm = buf.trim().parse::<i8>()?;
+        if perm != -1 {
+            let msg = format!("Tracing not permitted: you must be root or {} must contain -1",
+                           PERF_PERMS_PATH);
+            return Err(TraceMeError::TracingNotPermitted(msg));
+        }
+
+        Ok(())
+    }
+}
+
+impl Tracer for PerfPTTracer {
+    fn start_tracing(&mut self) -> Result<(), TraceMeError> {
+        PerfPTTracer::check_perf_perms()?;
+        if self.tracer_ctx.is_some() {
```

Comment:
Shouldn't this `if` be a statement ending in `?`?

Reply:
You mean using an `and*` combinator or something?

Reply:
Yes, something like that.

Reply:
I'm not sure these help actually. We have:

```
fn and<U>(self, optb: Option<U>) -> Option<U>[src]

Returns None if the option is None, otherwise returns optb.
```

 and

```
fn and_then<U, F>(self, f: F) -> Option<U> where

Returns None if the option is None, otherwise calls f with the wrapped value and returns the result.
```

We would need something which returns a `Result`.

We do have:
```
fn ok_or<E>(self, err: E) -> Result<T, E>[src]

Transforms the Option<T> into a Result<T, E>, mapping Some(v) to Ok(v) and None to Err(err).
```

But we would want the opposite mapping.

Am I missing something?

Reply:
Ignore this comment -- I was being an idiot. Mea culpa!